File: leetcode/c++/138_copy_list_with_random_pointer.cpp
// https://leetcode.cn/problems/copy-list-with-random-pointer/description

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*, Node*> hashArray;
        Node* newHead1 = head;
        Node* newHead2 = head;
        while (newHead1 != NULL) {
            hashArray[newHead1] = new Node(newHead1->val);
            newHead1 = newHead1->next;
        }
        while (newHead2 != NULL) {
            hashArray[newHead2]->next = hashArray[newHead2->next];
            hashArray[newHead2]->random = hashArray[newHead2->random];
            newHead2 = newHead2->next;
        }
        return hashArray[head];
    }
};

// 时间复杂度和空间复杂度均为O(n)

File: leetcode/golang/138_copy_list_with_random_pointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

 func copyRandomList(head *Node) *Node {
    newhead1 , newhead2 := head, head
    hashArray := make(map[*Node]*Node)
    for newhead1 != nil {
        hashArray[newhead1] = &Node {
            Val: newhead1.Val,
        }
        newhead1 = newhead1.Next
    }
    for newhead2 != nil {
        hashArray[newhead2].Next = hashArray[newhead2.Next] 
        hashArray[newhead2].Random = hashArray[newhead2.Random] 
        newhead2 = newhead2.Next
    }
    return hashArray[head]
}